### PR TITLE
Added error styling, and associated test, to change password page

### PR DESCRIPTION
**Pull request by GDSNewt**
Updated the styling on the Change Password page to be more consistent with the rest of GOV.UK. 

**Old styling**
Screen_Shot_2019-11-18_at_15 46 28

**New styling**
Screen Shot 2019-12-11 at 14 56 24

### CODE REVIEW

File: app/views/password/password_form.html.erb [hunk 1]
```diff
@@ -4,18 +4,24 @@
...
 <h1 class="govuk-heading-l"><%= t 'password.change_password_heading' %></h1>
 
-<%= form_for @password_form, url: profile_change_password_path do |f| %>
+<%=error_summary_for(@password_form&.errors, :form) %>
+<%= form_for @password_form, url: profile_change_password_path, as: :form do |f| %>
   <fieldset class="govuk-fieldset" aria-describedby="changed-password-hint">
-    <div class="govuk-form-group">
-      <%= f.label :old_password, t('password.current_password_lbl'), class: "govuk-label" %>
-      <%= f.password_field :old_password, class: "govuk-input govuk-input--width-20" %>
+    <div class="govuk-form-group <%= 'govuk-form-group--error' if @form&.errors&.key?(:old_password) %>">
```
Comment (j-maynard) [inline app/views/password/password_form.html.erb]: your @form object isn't an object and is nil....  thank @jakubmiarka for this...  You need to use `@password_form.errors.key?(old_password)`... This gets rid of the & and repeat for lines 16 and 22.

Comment (GDSNewt) [reply app/views/password/password_form.html.erb]: Updated, please re-review 🙏

File: app/views/password/password_form.html.erb [hunk 2]
```diff
@@ -4,20 +4,26 @@
...
 
-<%= form_for @password_form, url: profile_change_password_path do |f| %>
+<%=error_summary_for(@password_form&.errors, :form) %>
+<%= form_for @password_form, url: profile_change_password_path, as: :form do |f| %>
   <fieldset class="govuk-fieldset" aria-describedby="changed-password-hint">
-    <div class="govuk-form-group">
-      <%= f.label :old_password, t('password.current_password_lbl'), class: "govuk-label" %>
-      <%= f.password_field :old_password, class: "govuk-input govuk-input--width-20" %>
+    <div class="govuk-form-group <%= 'govuk-form-group--error' if @password_form.errors.key?(:old_password) %>">
+      <h3 class="govuk-heading-m"><%= f.label :old_password, t('password.old_password_lbl'), class: "govuk-label" %></h3>
```
Comment (jakubmiarka) [inline app/views/password/password_form.html.erb]: these shouldn't be headings but labels, and be as per the design system to ensure accessibility and consistency

File: app/views/password/password_form.html.erb [hunk 3]
```diff
@@ -4,20 +4,26 @@
 
 <h1 class="govuk-heading-l"><%= t 'password.change_password_heading' %></h1>
 
-<%= form_for @password_form, url: profile_change_password_path do |f| %>
+<%=error_summary_for(@password_form&.errors, :form) %>
```
Comment (jakubmiarka) [inline app/views/password/password_form.html.erb]:
```suggestion
<%= error_summary_for(@password_form.errors, :form) %>
```

File: config/locales/en.yml [hunk 4]
```diff
@@ -199,9 +199,9 @@ en:
     reset_password_heading: Set your new password
     password_changed: Password changed successfully
     reset_password_legend: You can reset your password by entering the code you have received via email.
+    old_password_lbl: Old password
```
Comment (jakubmiarka) [inline config/locales/en.yml]: why is this being changed?

File: app/views/password/password_form.html.erb [hunk 5]
```diff
@@ -4,20 +4,26 @@
...
+<%= form_for @password_form, url: profile_change_password_path, as: :form do |f| %>
   <fieldset class="govuk-fieldset" aria-describedby="changed-password-hint">
-    <div class="govuk-form-group">
-      <%= f.label :old_password, t('password.current_password_lbl'), class: "govuk-label" %>
-      <%= f.password_field :old_password, class: "govuk-input govuk-input--width-20" %>
+    <div class="govuk-form-group <%= 'govuk-form-group--error' if @password_form.errors.key?(:old_password) %>">
+      <h3 class="govuk-heading-m"><%= f.label :old_password, t('password.old_password_lbl'), class: "govuk-label" %></h3>
+      <%= error_message_on(f.object.errors, :old_password) %>
+      <%= f.password_field :old_password, autofocus: true, class: "govuk-input govuk-input--width-20" %>
```
Comment (jakubmiarka) [inline app/views/password/password_form.html.erb]: the field should have a conditional class `govuk-input--error` to add the red border around it

Comment (jakubmiarka) [reply app/views/password/password_form.html.erb]: once fixed, update the screenshot in the PR please

File: spec/system/visit_change_password_spec.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+require 'rails_helper'
+
+RSpec.describe 'ChangePassword', type: :system do
+  before(:each) do
+    login_gds_user
+  end
+
+  it 'shows form' do
+    visit profile_change_password_path()
```
Comment (jakubmiarka) [inline spec/system/visit_change_password_spec.rb]:
```suggestion
    visit profile_change_password_path
```

File: spec/system/visit_change_password_spec.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    expect(page).to have_content 'Old password'
+  end
+
+  it 'shows errors when an empty form is submitted' do
+    visit profile_change_password_path()
```
Comment (jakubmiarka) [inline spec/system/visit_change_password_spec.rb]:
```suggestion
    visit profile_change_password_path
```

File: spec/system/visit_change_password_spec.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
+require 'rails_helper'
+
+RSpec.describe 'ChangePassword', type: :system do
+  before(:each) do
+    login_gds_user
+  end
+
+  it 'shows form' do
+    visit profile_change_password_path()
+    expect(page).to have_content 'Old password'
```
Comment (jakubmiarka) [inline spec/system/visit_change_password_spec.rb]: the text shouldn't be hardcoded but use the locale instead, e.g. `t('password.change_password_btn')`

File: spec/system/visit_change_password_spec.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  it 'shows errors when an empty form is submitted' do
+    visit profile_change_password_path()
+    click_button 'Change password'
+    forms = page.find("#new_form > fieldset")
+    expect(forms).to have_content "Old password can't be blank"
+    expect(forms).to have_content "Password can't be blank"
+    expect(forms).to have_content "Password confirmation can't be blank"
+  end
+end
+
```
Comment (jakubmiarka) [inline spec/system/visit_change_password_spec.rb]: you should have a happy path test.. your changes broke the change password functionality...

File: spec/system/visit_change_password_spec.rb (hunk 8 above)
Comment (Rosa-Fox) [inline spec/system/visit_change_password_spec.rb]: I would probably check for the heading 'Change your password'

File: app/views/password/password_form.html.erb [hunk 10]
```diff
@@ -4,20 +4,26 @@
 
 <h1 class="govuk-heading-l"><%= t 'password.change_password_heading' %></h1>
 
-<%= form_for @password_form, url: profile_change_password_path do |f| %>
+<%=error_summary_for(@password_form&.errors, :form) %>
+<%= form_for @password_form, url: profile_change_password_path, as: :form do |f| %>
```
Comment (jakubmiarka) [inline app/views/password/password_form.html.erb]: by adding the `as: :form` you're now sending the changes to the controller under the `form` params (`params[:form]`), but the controller expects the content be in  the `change_password_form` params (`params[:change_password_form]`)

File: spec/system/visit_change_password_spec.rb [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  it 'shows form' do
+    visit profile_change_password_path()
+    expect(page).to have_content 'Old password'
+  end
+
+  it 'shows errors when an empty form is submitted' do
+    visit profile_change_password_path()
+    click_button 'Change password'
+    forms = page.find("#new_form > fieldset")
+    expect(forms).to have_content "Old password can't be blank"
```
Comment (Rosa-Fox) [inline spec/system/visit_change_password_spec.rb]: Can these be localised?

File: config/locales/en.yml (hunk 4 above)
Comment (GDSNewt) [reply config/locales/en.yml]: Screen Shot 2019-12-10 at 16 55 42

I think I changed it to be more consistent with the error messages associated with password. But I'm not sure if that is a strong enough rationale so I have changed it back

Comment (jakubmiarka) [reply config/locales/en.yml]: The content has been signed off by content people and everything should be as per the protype, so let's not make any content changes...

File: spec/system/visit_change_password_spec.rb (hunk 11 above)
Comment (GDSNewt) [reply spec/system/visit_change_password_spec.rb]: I'm not sure. The errors don't utilise yml as far I can tell. I've tried looking at how the error messages are generated but i'm not really sure how it works 😕

File: spec/system/visit_change_password_spec.rb [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    visit profile_change_password_path
+    click_button 'Change password'
+    forms = page.find("#new_change_password_form")
+    expect(forms).to have_content "Old password can't be blank"
+    expect(forms).to have_content "Password can't be blank"
+    expect(forms).to have_content "Password confirmation can't be blank"
+  end
+
+  it 'submits the form correctly without error' do 
+    stub_cognito_response(method: :change_password,payload: {})
```
Comment (GDSNewt) [inline spec/system/visit_change_password_spec.rb]: Using the stub cognito response here helped stop my pre commit tests from failing but i'm not sure i've used it entirely correctly here, would appreciate some feedback on this 👍

Comment (jakubmiarka) [reply spec/system/visit_change_password_spec.rb]: yep, this is perfectly fine

**Pull request merged**